util/util_tls.go

```diff
@@ -0,0 +1,171 @@
+		ctx.WithFields(log.Fields{
+			"host":     host,
+			"certPath": certPath,
+			"certKey":  certPath,
+		}).Debug("skipping self-cert creation, files exist")
+		return nil
+	}
+
+	certRoot := filepath.Dir(certPath)
+	keyRoot := filepath.Dir(keyPath)
+	if err := os.MkdirAll(certRoot, 0755); err != nil {
+		ctx.WithFields(log.Fields{
+			"host":     host,
+			"certRoot": certRoot,
+		}).Debug("created dir")
+
+		return err
+	}
+	if keyRoot != certRoot {
+		if err := os.MkdirAll(keyRoot, 0755); err != nil {
+			ctx.WithFields(log.Fields{
+				"host":    host,
+				"keyRoot": keyRoot,
+			}).Debug("created dir")
+			return err
+		}
+	}
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return err
+	}
+
+	tmpl := x509.Certificate{
+		SerialNumber:       serialNumber,
+		Subject:            pkix.Name{Organization: []string{orgName}, CommonName: host},
+		NotBefore:          time.Now(),
+		NotAfter:           time.Now().AddDate(1, 0, 0),
+		SignatureAlgorithm: x509.SHA256WithRSA,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	if ip := net.ParseIP(host); ip != nil {
+		ctx.WithField("ip", ip).Debug("ip used in cert template")
+		tmpl.IPAddresses = append(tmpl.IPAddresses, ip)
+	} else {
+		ctx.WithField("host", host).Debug("host used in cert template")
+		tmpl.DNSNames = append(tmpl.DNSNames, host)
+	}
+
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	// gen cert file
+	ctx.WithField("certFile", certPath).Debug("creating cert file")
+	certBlock, err := x509.CreateCertificate(rand.Reader, &tmpl, &tmpl, &privKey.PublicKey, privKey)
+	if err != nil {
+		return err
+	}
+
+	certFile, err := os.Create(certPath)
+	if err != nil {
+		return err
+	}
+	defer certFile.Close()
+	if err := pem.Encode(certFile, &pem.Block{Type: certBlockType, Bytes: certBlock}); err != nil {
+		return err
+	}
+
+	// gen key file
+	ctx.WithField("keyFile", keyPath).Debug("creating key file")
+	keyFile, err := os.OpenFile(keyPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return err
+	}
+	defer keyFile.Close()
+
+	keyBlock := x509.MarshalPKCS1PrivateKey(privKey)
+	if err != nil {
+		return err
+	}
+
+	if err := pem.Encode(keyFile, &pem.Block{Type: keyBlockType, Bytes: keyBlock}); err != nil {
+		return err
+	}
+
+	ctx.WithFields(log.Fields{
+		"certPath": certPath,
+		"certKey":  certPath,
+	}).Debug("self-cert files created")
+
+	return nil
+}
+
+func AssertTrustedHost(ctx apitypes.Context, host, algo string, fingerprint []byte) bool {
+	trusted := "no"
+	fmt.Printf("Rejecting connection to unknown host %s.\n", host)
+	fmt.Printf("SHA Fingerprint presented: %s:%x.\n", algo, fingerprint[0:32])
+	fmt.Print("Do you want to save host to knwon_hosts file? (yes/no): ")
```

Comment:
Hi @vladimirvivien,

Oops, a small spelling error on the filename `kwon_hosts`.